Hw06

zuwarka opened the pull request.

zuwarka committed:
Create README.md

zuwarka committed:
Create main.rb

Работа классов VersionManager и BinaryTree

zuwarka committed:
Create binary_tree.rb

zuwarka committed:
Create node.rb

zuwarka committed:
Create version_manager.rb

Alex6Alex commented on homework06 (22.07.2022)/binary_tree.rb:
```diff
@@ -0,0 +1,91 @@
+    end
+  end
+
+  def find(value, node = self.root)
```
Методы `find`, `delete`, `to_s` стоит попробовать реализовать так, чтобы основная логика ушла в ноду, как это сделано с методов добавления нового элемента. Тогда не будет необходимости держать параметр `node = self.root`. Сейчас получается, что этот параметр можно передать извне и нарушить работу метода: `tree.find(1, nil)`, `tree.find(1, Node.new)`

Alex6Alex commented on homework06 (22.07.2022)/binary_tree.rb:
```diff
@@ -0,0 +1,91 @@
+
+  private
+  def find_min(node = self.root)
+    if node == nil
```
При написании условий можно избавляться от избыточных условий в пользу guard clause:
```
return if node.nil?

return node if node.left.nil?
find_min(node.left)
```

Alex6Alex commented on homework06 (22.07.2022)/version_manager.rb:
```diff
@@ -0,0 +1,66 @@
+# -- Если версия не указана, то значение по умолчанию - “0.0.1”.
+
+class String
+  def is_integer?
```
Исходные классы языка лучше не расширять лишний раз, если это не понадобится в проекте в большинстве мест. А так лучше сделать отдельный класс, который обрабатывает какую-то логику, если логика довольно большая

Alex6Alex commented on homework06 (22.07.2022)/version_manager.rb:
```diff
@@ -0,0 +1,66 @@
+  def initialize(m_version = "0.0.1")
+    @prev_version = m_version
+    check_integer
+    @major = m_version.split(/[.]/)[0].to_i
```
`m_version.split(/[.]/)` лучше сделать один раз и записать в локальную переменную. Иначе, три раз создается новый массив. Регулярное выражение тут тоже, пожалуй, лишнее. Можно проще: `split('.')`

Alex6Alex commented on homework06 (22.07.2022)/version_manager.rb:
```diff
@@ -0,0 +1,66 @@
+    if @prev_version == nil
+      raise 'Can not rollback to previous version'
+    end
+    @major = @prev_version.split(/[.]/)[0].to_i
```
Здесь тоже можно воспользоваться локальной переменной для хранения одного массива

zuwarka committed:
Update node.rb

Изменена логика ноды - теперь рекурсивные функции find и delete добавлены в ноду вместо дерева. Дерево будет только вызывать функции ноды.

zuwarka committed:
Update binary_tree.rb

Изменена логика дерева - теперь рекурсивные функции find и delete перенесены в ноду.

zuwarka committed:
Update version_manager.rb

Вместо расширения String прописан новый класс VersionValidator, в котором проверяем значения версии на целые числа и на, собственно, целые числа (Integer). В VersionManager'e создаем объект класса VersionValidator и вызываем метод check_integer.
Сделан рефакторинг метода rollback! - условие ошибки (строка 58) записано в одну строку.

zuwarka committed:
Update version_manager.rb

В конструкторе VersionManager переменные @major, @minor, @patch объявляются через @prev_version.
